src/coreclr/jit/importer.cpp

```diff
@@ -2911,6 +2911,27 @@ int Compiler::impBoxPatternMatch(CORINFO_RESOLVED_TOKEN* pResolvedToken,
         case CEE_ISINST:
             if (codeAddr + 1 + sizeof(mdToken) + 1 <= codeEndp)
             {
+                // First, let's see if we can fold BOX+ISINST to just null if ISINST is known to return null
+                // for the given argument. Don't make inline observations for this case.
+                if ((opts != BoxPatterns::MakeInlineObservation) &&
+                    ((impStackTop().val->gtFlags & GTF_SIDE_EFFECT) == 0))
+                {
+                    if ((opts == BoxPatterns::IsByRefLike) ||
```

Comment:
> A separate question -- doesn't this logic essentially subsume the logic that happens below?

Sort of, in this case we only handle MustNot case while the logic below handles both Must and MustNot - I tried to share them but decide that it's not worth it, also, removed nullable<> support since it didn't add any diffs

> Should we do this for byref-like types? 
Let me remove it then and keep only the patterns listed there then